ouroboros-network/src/Ouroboros/Network/Connections/Types.hs

```diff
@@ -0,0 +1,178 @@
+-- can transport data in `m`.
+--
+-- For `Existing` resources, if the decision is `Rejected`, or if the call
+-- throws an exception (`m` is IO-capable), then the caller is responsible for
+-- closing the resource. Otherwise, the `Connections` term must ensure it is
+-- closed appropriately.
+--
+-- For `New` resources, the caller does not create the resource, so the
+-- `Connections` term is responsible for creating and closing it appropriately.
+--
+-- Note on exceptions: in case `m` can throw exceptions, there's some
+-- ambiguity about when it's a good idea to do so, because it's not clear
+-- whether an exception should be caught and used to give a `Rejected` value,
+-- or simply ignore/re-thrown.
+-- If the resource acquisition of a `New` connection throws an exception, that
+-- should be re-thrown. This is consistent with bracketing of the acquire and
+-- release fields of the `New` constructor, which also re-throws the exception.
+--
+-- TODO may need a way to check / be notified when there is an incoming
+-- connection at a given identifier.
+-- Could just put that on the concurrent connections: an `STM (Map id handle)`.
+-- When you get the handle back, the underlying connection and its handler
+-- thread may of course have died, but that's always a possibility even if
+-- you `include`. 
+newtype Connections id resource request reject accept m = Connections
+  { include :: forall provenance .
+               id
+            -> Resource provenance m resource
+            -> request provenance
+            -> m (Decision provenance reject accept)
+  }
+
+-- | Connections is a contravariant functor in request, but sadly it's not the
+-- final type parameter so we can't give a Contravariant instance. Also, the
+-- request type is not of kind `Type` so it wouldn't work anyway.
+contramapRequest
+  :: (forall provenance . req1 provenance -> req2 provenance)
+  -> Connections id resource req2 reject accept m
+  -> Connections id resource req1 reject accept m
+contramapRequest f connections = connections
+  { include = \ident res req -> include connections ident res (f req) }
+
+mapResult
+  :: ( Functor m )
+  => (forall provenance . reject1 provenance -> reject2 provenance)
+  -> (forall provenance . accept1 provenance -> accept2 provenance)
+  -> Connections id resource req reject1 accept1 m
+  -> Connections id resource req reject2 accept2 m
+mapResult frej facc connections = connections
+  { include = \ident res req -> fmap (mapDecision frej facc) (include connections ident res req) }
+
+-- Client and Server types take any suitable Connections include callback,
+-- specialized to Outgoing or Incoming, and use it to get a decision, for any
+-- accept or reject type.
+
+-- | A `Client` takes a continuation that is derived from a `Connections`
+-- (see `runClientWith`) and uses it to get a decision parameteric in accept
+-- and reject.
+type Client addr sock m request = forall accept reject .
```

Comment:
I would rename this `Client` type to `ClientSocket` or something similar. The construction of `Client` instruments how to construct a socket for a client connection - I would also put such a sentence to the haddock, and link add a reference to `Connections.Socket.Client.client`.

Reply:
How about `RunClient` or `MkClient`?